### PR TITLE
80738  Get IPOs on project with filter and sorting

### PR DESCRIPTION
**Pull request by bratli**
DevOps#80738

### CODE REVIEW

File: src/Equinor.ProCoSys.IPO.Domain/DateTimeExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace Equinor.ProCoSys.IPO.Domain
+{
+    public static class DateTimeExtensions
+    {
+        private static readonly DayOfWeek firstDayOfWeek = DayOfWeek.Monday;
+
+        public static DateTime AddWeeks(this DateTime dateTime, int weeks)
+            => dateTime.AddDays(7 * weeks);
+
+        public static DateTime StartOfWeek(this DateTime dt)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/DateTimeExtensions.cs:12]: Add 3 unit tests. Pick a time on a Monday and Sunday (these tests exists in pres)

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                            invitation.CommPkgs.Any(c => c.CommPkgNo.ToUpper().StartsWith(filter.CommPkgNoStartsWith.ToUpper())) ||
+                            invitation.McPkgs.Any(mc => mc.CommPkgNo.ToUpper().StartsWith(filter.CommPkgNoStartsWith.ToUpper()))) &&
+                      (string.IsNullOrEmpty(filter.McPkgNoStartsWith) ||
+                            invitation.McPkgs.Any(mc => mc.McPkgNo.ToUpper().StartsWith(filter.McPkgNoStartsWith.ToUpper()))) &&
+                      (string.IsNullOrEmpty(filter.TitleStartsWith) ||
+                            invitation.Title.ToUpper().StartsWith(filter.TitleStartsWith.ToUpper())) &&
+                      (filter.PersonOid == null ||
+                            invitation.Participants.Any(p => p.AzureOid == filter.PersonOid)) &&
+                      (filter.FunctionalRoleCode == null ||
+                       invitation.Participants.Any(p => p.FunctionalRoleCode.ToUpper() == filter.FunctionalRoleCode.ToUpper())) &&
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:51]: indent

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                            invitation.StartTimeUtc >= filter.PunchOutDateFromUtc) &&
+                      (filter.PunchOutDateToUtc == null ||
+                            invitation.EndTimeUtc <= filter.PunchOutDateToUtc) &&
+                      (filter.LastChangedAtFromUtc == null ||
+                            (invitation.ModifiedAtUtc ?? invitation.CreatedAtUtc) >= filter.LastChangedAtFromUtc) &&
+                      (filter.LastChangedAtToUtc == null ||
+                            (invitation.ModifiedAtUtc ?? invitation.CreatedAtUtc) <= filter.LastChangedAtToUtc)
+                select new InvitationDto
+                {
+                    Id = EF.Property<int>(invitation, "Id"),
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:62]: why not invitation.Id here?

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate;
+using Microsoft.EntityFrameworkCore;
+
+namespace Equinor.ProCoSys.IPO.Query.GetInvitations
+{
+    public abstract class GetInvitationsQueryBase
+    {
+        protected IQueryable<InvitationDto> CreateQueryableWithFilter(IReadOnlyContext context, string projectName, Filter filter, DateTime utcNow)
+        {
+            var startOfThisWeekUtc = DateTime.MinValue;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:14]: So the business logic for when an invitation is Overdue in IPO follow the same rules as preservation?
In preservation it is OK to preserve within the week. It is ok to preserve on a Tuesday and on the Sunday. Same same.

So: If the IPO round is invited to happen on Tuesday at 14:30, is it overdue when that paricular week has ended (I.e 1 second after midnight at upcoming Sunday)

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                    Status = invitation.Status,
+                    Type = invitation.Type,
+                    CreatedAtUtc = invitation.CreatedAtUtc,
+                    StartTimeUtc = invitation.StartTimeUtc,
+                    EndTimeUtc = invitation.EndTimeUtc,
+                    CompletedAtUtc = invitation.CompletedAtUtc,
+                    AcceptedAtUtc = invitation.AcceptedAtUtc,
+                    ContractorRep = GetContractorRep(invitation.Participants.ToList()),
+                    ConstructionCompanyRep = GetConstructionCompanyRep(invitation.Participants.ToList()),
+                    McPkgNos = invitation.McPkgs.Select(mc => mc.McPkgNo).ToList(),
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:75]: I really don't understand how the McPkgs/CommPgs are fetched from DB here ... when seeing code, I was pretty sure that both list of McPkgs/CommPgs was empty after quering. But no, it works and I see that U cover it in tests.

Du you understand the magic here? Do .Select(mc => mc.McPkgNo) actually build a join in the executed query?

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs [hunk 6]
```diff
@@ -38,6 +40,22 @@ public class InvitationsController : ControllerBase
 
         public InvitationsController(IMediator mediator) => _mediator = mediator;
 
+        [Authorize(Roles = Permissions.IPO_READ)]
+        [HttpGet]
+        public async Task<ActionResult<InvitationsResult>> GetTags(
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs:45]: GetInvitations

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+                var query = new GetInvitationsQuery(_projectName2, null, filter);
+                var dut = new GetInvitationsQueryHandler(context);
+
+                var result = await dut.Handle(query, default);
+                AssertCount(result.Data, 1);
+            }
+        }
+
+        [TestMethod]
+        public async Task Handler_ShouldFilterOnCommPkgNoStartWith2()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs:417]: rename Handler_ShouldFilterOnCommPkgNoStartWith2 .. I guess you test that the commpkg is found via McPkg and the testname should highlight so (if I'm right)

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+                var query = new GetInvitationsQuery(_projectName2, null, filter);
+                var dut = new GetInvitationsQueryHandler(context);
+
+                var result = await dut.Handle(query, default);
+                AssertCount(result.Data, 2);
+            }
+        }
+
+        [TestMethod]
+        public async Task Handler_ShouldFilterOnPunchOutInNextWeek()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs:465]: Consider to duplicate this test, but use TimeService to move the time backward so filtering on NextWeek can end with AssertCount(result.Data, 1);

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+                var query = new GetInvitationsQuery(_projectName2, null, filter);
+                var dut = new GetInvitationsQueryHandler(context);
+
+                var result = await dut.Handle(query, default);
+                AssertCount(result.Data, 2);
+            }
+        }
+
+        [TestMethod]
+        public async Task Handler_ShouldFilterOnLastChangedAtFrom2()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs:497]: rename Handler_ShouldFilterOnLastChangedAtFrom2 so reader understand why expecting result is as it is

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs [hunk 10]
```diff
@@ -0,0 +1,582 @@
...
+                var query = new GetInvitationsQuery(_projectName2, null, filter);
+                var dut = new GetInvitationsQueryHandler(context);
+
+                var result = await dut.Handle(query, default);
+                AssertCount(result.Data, 0);
+            }
+        }
+
+        [TestMethod]
+        public async Task Handler_ShouldFilterOnLastChangedAtTo2()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs:529]: rename Handler_ShouldFilterOnLastChangedAtTo2 so reader understand why expecting result is as it is

**Review by janid1967 — APPROVED** [review]
Approved with suggestions

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+                startOfNextWeekUtc = startOfThisWeekUtc.AddWeeks(1);
+                startOfTwoWeeksUtc = startOfThisWeekUtc.AddWeeks(2);
+            }
+
+            var ipoIdStartWith = GetIpoIdStartWith(filter.IpoIdStartsWith);
+ 
+            var queryable = from invitation in context.QuerySet<Invitation>()
+                where invitation.ProjectName == projectName &&
+                      (!filter.PunchOutDates.Any() ||
+                           (filter.PunchOutDates.Contains(PunchOutDateFilterType.Overdue) && invitation.StartTimeUtc < startOfThisWeekUtc) ||
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:29]: If the IPO round is invited to happen on Tuesday at 14:30, is it overdue already when the particular week starts?
Isn't overdue when invitation.StartTimeUtc > utcNow?

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs (hunk 5 above)
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:75]: I do not know the magic behind this, actually expected it to be empty myself, and that I needed to have .Include for all of these, but it works as is.

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs (hunk 11 above)
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:29]: Yes, true, will fix this!

File: src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs (hunk 3 above)
Comment (bratli) [reply src/Equinor.ProCoSys.IPO.Query/GetInvitations/GetInvitationsQueryBase.cs:62]: At one stage of my query it actually didnt work, but now it works I see (with invitation.Id)

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs [hunk 12]
```diff
@@ -477,6 +478,22 @@ public async Task Handler_ShouldFilterOnPunchOutInNextWeek()
             }
         }
 
+        [TestMethod]
+        public async Task Handler_ShouldFilterOnPunchOutInNextWeek_ShouldReturnInvitation()
+        {
+            TimeService.SetProvider(new ManualTimeProvider(new DateTime(2019, 10, 25, 11, 0, 0, DateTimeKind.Utc)));
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitations/GetInvitationsQueryHandlerTests.cs:484]: You already have the ManualTimeProvider from base-class.
Delete this line and use _timeProvider.ElapseWeeks(N); where N is number of weeks. I guess you should have -1

**Pull request merged**